### PR TITLE
User Authentication with BCRYPT and RSPEC

### PR DESCRIPTION
**Pull request by Eunicegenel**
PR Notes
- Created user model that has username, email, password ( hashed with BCrypt ), role, status and money.
- Created session model responsible for being the overall dashboard for admin, buyers and brokers
- Implemented dashboard division ( there are 3 different dashboards for admin, buyers and brokers )
- User creation, user authentication and user logouts are all fully functioning
- RSPECs made for models, requests and features are good and correct ( user model and controller only )
- RSPEC features for buyer registration, broker registration, buyer/admin login, broker pending login, broker approved login and user logout are complete and good

Integrated User Auth RSPEC
image

Future Actions
- Proceeding with mailer implementation
- After mailer comes admin pending functionality

### CODE REVIEW

File: .github/workflows/rspec.yml [hunk 1]
```diff
@@ -49,4 +49,3 @@ jobs:
         DATABASE_HOST: 127.0.0.1
         DATABASE_USERNAME: 'postgres'
         DATABASE_PASSWORD: 'test'
-        RAILS_MASTER_KEY: ${{ secrets.RAILS_MASTER_KEY }}
```
Comment (analyn-cajocson) [inline .github/workflows/rspec.yml:52]: Why did you delete this key? Make sure that you will not be using the `credentials.yml.enc` for hiding your API token

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+  validates :email, presence: true, uniqueness: true
+  validates :username, presence: true, uniqueness: true
+  validates :role, presence: true
+  # validates_inclusion_of :role, :in => %w(Buyer Broker Admin)
```
Comment (analyn-cajocson) [inline app/models/user.rb]: Remove the unnecessary comments

File: config/routes.rb [hunk 3]
```diff
@@ -1,3 +1,11 @@
...
+  resources :users, only: [:new, :create]
+  root :to => 'sessions#welcome'
+    get 'login', to: 'sessions#new'
+  post 'login', to: 'sessions#create'
+  get 'welcome', to: 'sessions#welcome'
+  get 'dashboard_buyer', to: 'sessions#index_buyer'
+  get 'dashboard_broker', to: 'sessions#index_broker'
+  get 'dashboard_admin', to: 'sessions#index_admin'
+  get 'logout', to: 'sessions#out'
+end
```
Comment (analyn-cajocson) [inline config/routes.rb]: Add new line at the end of file

File: config/routes.rb [hunk 4]
```diff
@@ -1,3 +1,11 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
-end
+  resources :users, only: [:new, :create]
+  root :to => 'sessions#welcome'
+    get 'login', to: 'sessions#new'
```
Comment (analyn-cajocson) [inline config/routes.rb]: Fix code alignment

**Review by analyn-cajocson — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 4 above)
Comment (analyn-cajocson) [reply config/routes.rb]: @Eunicegenel

**Review by analyn-cajocson — CHANGES_REQUESTED** [review]

**Review by analyn-cajocson — APPROVED** [review]

**Review by lnabrigo18 — APPROVED** [review]
